File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/RowExamples.java
package questpdf.docexamples;

import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSize;
import com.questpdf.infrastructure.IContainer;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;

public class RowExamples extends DocExample {

    @Test
    public void simpleExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.margin(25f);

                    page.content()
                        .padding(25f)
                        .width(325f)
                        .row(row -> {
                            row.constantItem(100f)
                                .background(Colors.Grey.getMedium())
                                .padding(10f)
                                .text("100pt");

                            row.relativeItem()
                                .background(Colors.Grey.getLighten1())
                                .padding(10f)
                                .text("75pt");

                            row.relativeItem(2f)
                                .background(Colors.Grey.getLighten2())
                                .padding(10f)
                                .text("150pt");
                        });
                });
            })
            .generateImages(index -> output("row-simple.webp"), settings);
    }

    @Test
    public void spacingExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1000f));
                    page.margin(25f);

                    page.content()
                        .padding(25f)
                        .width(220f)
                        .height(50f)
                        .row(row -> {
                            row.spacing(10f);

                            row.relativeItem(2f).background(Colors.Grey.getMedium());
                            row.relativeItem(3f).background(Colors.Grey.getLighten1());
                            row.relativeItem(5f).background(Colors.Grey.getLighten2());
                        });
                });
            })
            .generateImages(index -> output("row-spacing.webp"), settings);
    }

    @Test
    public void customSpacingExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(250f, 0f));
                    page.maxSize(new PageSize(250f, 1000f));
                    page.margin(25f);

                    page.content()
                        .height(50f)
                        .row(row -> {
                            row.relativeItem().background(Colors.Grey.getDarken1());
                            row.constantItem(10f);
                            row.relativeItem().background(Colors.Grey.getMedium());
                            row.constantItem(20f);
                            row.relativeItem().background(Colors.Grey.getLighten1());
                            row.constantItem(30f);
                            row.relativeItem().background(Colors.Grey.getLighten2());
                        });
                });
            })
            .generateImages(index -> output("row-spacing-custom.webp"), settings);
    }

    @Test
    public void disableUniformItemsHeightExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.High);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(700f, 0f));
                    page.maxSize(new PageSize(700f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);
                    page.pageColor(Colors.getWhite());

                    page.content()
                        .row(row -> {
                            row.spacing(15f);

                            labelStyle(row.relativeItem())
                                .text("Programming is both a science and an art — it demands precision, creativity, and patience. At its core, it’s about understanding how to break down complex problems into small, logical steps that a computer can execute.");

                            labelStyle(row.relativeItem())
                                .text("Programming is the art of turning ideas into logic, logic into code, and code into something that solves real problems.");
                        });
                });
            })
            .generateImages(index -> output("row-uniform-height-enabled.webp"), settings);
    }

    // The C# original routes LabelStyle through the Element(Func<IContainer, IContainer>)
    // overload, which is not bridged; a static helper method is the direct equivalent.
    private static IContainer labelStyle(IContainer container) {
        return container
            .shrinkVertical()
            .background(Colors.Grey.getLighten3())
            .cornerRadius(15f)
            .padding(15f);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/RowExamples.kt
package questpdf.docexamples

import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSize
import com.questpdf.infrastructure.IContainer
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test

class RowExamples : DocExample() {

    @Test
    fun simpleExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    margin(25f)

                    content()
                        .padding(25f)
                        .width(325f)
                        .row {
                            constantItem(100f)
                                .background(Colors.Grey.Medium)
                                .padding(10f)
                                .text("100pt")

                            relativeItem()
                                .background(Colors.Grey.Lighten1)
                                .padding(10f)
                                .text("75pt")

                            relativeItem(2f)
                                .background(Colors.Grey.Lighten2)
                                .padding(10f)
                                .text("150pt")
                        }
                }
            }
            .generateImages({ output("row-simple.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun spacingExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1000f))
                    margin(25f)

                    content()
                        .padding(25f)
                        .width(220f)
                        .height(50f)
                        .row {
                            spacing(10f)

                            relativeItem(2f).background(Colors.Grey.Medium)
                            relativeItem(3f).background(Colors.Grey.Lighten1)
                            relativeItem(5f).background(Colors.Grey.Lighten2)
                        }
                }
            }
            .generateImages({ output("row-spacing.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun customSpacingExample() {
        Document
            .create {
                page {
                    minSize(PageSize(250f, 0f))
                    maxSize(PageSize(250f, 1000f))
                    margin(25f)

                    content()
                        .height(50f)
                        .row {
                            relativeItem().background(Colors.Grey.Darken1)
                            constantItem(10f)
                            relativeItem().background(Colors.Grey.Medium)
                            constantItem(20f)
                            relativeItem().background(Colors.Grey.Lighten1)
                            constantItem(30f)
                            relativeItem().background(Colors.Grey.Lighten2)
                        }
                }
            }
            .generateImages({ output("row-spacing-custom.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun disableUniformItemsHeightExample() {
        Document
            .create {
                page {
                    minSize(PageSize(700f, 0f))
                    maxSize(PageSize(700f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)
                    pageColor(Colors.White)

                    content()
                        .row {
                            spacing(15f)

                            // The C# original routes LabelStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; an extension function is the direct equivalent.
                            fun IContainer.labelStyle(): IContainer =
                                shrinkVertical()
                                    .background(Colors.Grey.Lighten3)
                                    .cornerRadius(15f)
                                    .padding(15f)

                            relativeItem()
                                .labelStyle()
                                .text("Programming is both a science and an art — it demands precision, creativity, and patience. At its core, it’s about understanding how to break down complex problems into small, logical steps that a computer can execute.")

                            relativeItem()
                                .labelStyle()
                                .text("Programming is the art of turning ideas into logic, logic into code, and code into something that solves real problems.")
                        }
                }
            }
            .generateImages({ output("row-uniform-height-enabled.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.High
                rasterDpi = 144
            })
    }
}
